plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-products-controller.php

```diff
@@ -504,32 +532,69 @@ public function add_search_criteria_to_wp_query_join( $join ) {
 	public function add_search_criteria_to_wp_query_where( $where ) {
 		global $wpdb;
 
-		if ( $this->search_name_or_sku_tokens ) {
-			$use_sku                  = wc_product_sku_enabled();
-			$posts_clause_parts       = array();
-			$meta_lookup_clause_parts = array();
-			foreach ( $this->search_name_or_sku_tokens as $token ) {
-				$like_search          = '%' . $wpdb->esc_like( $token ) . '%';
-				$posts_clause_parts[] = $wpdb->prepare( "($wpdb->posts.post_title LIKE %s)", $like_search );
-				if ( $use_sku ) {
-					$meta_lookup_clause_parts[] = $wpdb->prepare( '(wc_product_meta_lookup.sku LIKE %s)', $like_search );
-				}
-			}
-			$post_clause = implode( ' AND ', $posts_clause_parts );
-			if ( $use_sku ) {
-				$meta_lookup_clause = implode( ' AND ', $meta_lookup_clause_parts );
-			}
-			$where .=
-				$use_sku ?
-					" AND (($post_clause) OR ($meta_lookup_clause))" :
-					" AND ($post_clause)";
+		if ( $this->search_fields_tokens ) {
+			$where .= $this->build_dynamic_search_clauses(
+				$this->search_fields_tokens['tokens'],
+				$this->search_fields_tokens['fields']
+			);
+		} elseif ( $this->search_name_or_sku_tokens ) {
+			$searchable_fields = wc_product_sku_enabled() ? array( 'name', 'sku' ) : array( 'name' );
+			$where            .= $this->build_dynamic_search_clauses(
+				$this->search_name_or_sku_tokens,
+				$searchable_fields
+			);
 		} elseif ( ! empty( $this->search_sku_arg_value ) ) {
 			$like_search = '%' . $wpdb->esc_like( $this->search_sku_arg_value ) . '%';
 			$where      .= ' AND ' . $wpdb->prepare( '(wc_product_meta_lookup.sku LIKE %s)', $like_search );
 		}
 		return $where;
 	}
 
+	/**
+	 * Build search clauses for dynamic product search.
+	 *
+	 * @param array $tokens Search tokens.
+	 * @param array $fields Fields to search in.
+	 * @return string
+	 */
+	private function build_dynamic_search_clauses( $tokens, $fields ) {
+		global $wpdb;
+
+		if ( empty( $fields ) || empty( $tokens ) ) {
+			return '';
+		}
+
+		$column_map = array(
+			'name'              => "{$wpdb->posts}.post_title",
+			'sku'               => 'wc_product_meta_lookup.sku',
+			'global_unique_id'  => 'wc_product_meta_lookup.global_unique_id',
+			'description'       => "{$wpdb->posts}.post_content",
+			'short_description' => "{$wpdb->posts}.post_excerpt",
+		);
+
+		$field_clauses = array();
+
+		foreach ( $tokens as $token ) {
+			$like_search         = '%' . $wpdb->esc_like( $token ) . '%';
+			$field_token_clauses = array();
+
+			foreach ( $fields as $field ) {
+				if ( ! isset( $column_map[ $field ] ) ) {
+					continue;
+				}
+
+				$db_column             = $column_map[ $field ];
+				$field_token_clauses[] = '(' . $db_column . ' LIKE ' . $wpdb->prepare( '%s', $like_search ) . ')';
```

Comment:
Maybe something along these lines would be a tiny bit easier to grasp?

```suggestion
				$field_token_clauses[] = $wpdb->prepare( "({$db_column} LIKE %s)", $like_search );
```

Reply:
Thanks, definitely reads better 👍